server/src/main/java/org/tctalent/server/request/candidate/SearchCandidateRequest.java

```diff
@@ -152,6 +166,226 @@
+                ands.add("candidate_occupation.years_experience <= " + getMaxYrs());
+            }
+        }
+
+        // EXCLUDED CANDIDATES (eg from Review Status)
+        if (!ObjectUtils.isEmpty(excludedCandidates)) {
+            String values = excludedCandidates.stream()
+                .map(candidate -> candidate.getId().toString())
+                .collect(Collectors.joining(","));
+            ands.add("not candidate.id in (" + values + ")");
+        }
+
+        // NATIONALITY SEARCH
+        if (!ObjectUtils.isEmpty(getNationalityIds())) {
+            String values = getNationalityIds().stream()
+                .map(Objects::toString).collect(Collectors.joining(","));
+            if (getNationalitySearchType() == null || getNationalitySearchType().equals(SearchType.or)) {
+                ands.add("candidate.nationality_id in (" + values + ")");
+            } else {
+                ands.add("candidate.nationality_id not in (" + values + ")");
+            }
+        }
+
+        // COUNTRY SEARCH - taking into account user source country limitations
+        // If request ids is NOT EMPTY we can just accept them because the options
+        // presented to the user will be limited to the allowed source countries
+        if (!ObjectUtils.isEmpty(getCountryIds())) {
+            String values = getCountryIds().stream()
+                .map(Objects::toString).collect(Collectors.joining(","));
+            if (getCountrySearchType() == null || getCountrySearchType().equals(SearchType.or)) {
+                ands.add("candidate.country_id in (" + values + ")");
+            } else {
+                ands.add("candidate.country_id not in (" + values + ")");
+            }
+        // If request's countryIds IS EMPTY only show user's source countries
+        } else if (user != null && !ObjectUtils.isEmpty(user.getSourceCountries())) {
+            String values = user.getSourceCountries().stream()
+                .map(country -> country.getId().toString())
+                .collect(Collectors.joining(","));
+            ands.add("candidate.country_id in (" + values + ")");
+        }
+
+        // PARTNER SEARCH
+        if (!ObjectUtils.isEmpty(getPartnerIds())) {
+            joins.add(CandidateQueryHelper.CANDIDATE__USER__JOIN__NATIVE);
+            String values = getPartnerIds().stream()
+                .map(Objects::toString).collect(Collectors.joining(","));
+            ands.add("users.partner_id in (" + values + ")");
+        }
+
+        // SURVEY TYPE SEARCH
+        if (!ObjectUtils.isEmpty(getSurveyTypeIds())) {
+            String values = getSurveyTypeIds().stream()
+                .map(Objects::toString).collect(Collectors.joining(","));
+            ands.add("candidate.survey_type_id in (" + values + ")");
+        }
+
+        // REFERRER
+        final String referrerParam =
+            getRegoReferrerParam() == null ? null : getRegoReferrerParam().trim().toLowerCase();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

#### toLowerCase [](|2045d6a2-68fa-4cde-87a6-566ffddb2b76|)

The parameterless versions of `String.toLowerCase()` and `String.toUpperCase()` use the default locale of the JVM when transforming strings. This can have unintended results, including difficult-to-debug transient failures because case mapping differs based on locale. 

**Even if you aren’t writing code for a user experience that’s displayed to a customer**, system internal code often uses case folding to normalize strings for comparison.
Not doing so can missing locale lead to errors when running on systems that are configured to use a different system locale. This can include developer desktops or build servers (not just your fleet hosts): many hours have been spent debugging errors that only occur on a specific machine due to its locale configuration.

#### Recommended solutions:

Always pass in a `Locale`.

**Language or locale-dependent processing:**

Get the locale from the platform, use the language of the content, or configure it in your code and pass that to the method.
```
String firstName = &#34;Ichabod&#34;; 
Locale locale = // Where you obtain the locale from is platform dependent
String lowerCaseFirstName = firstName.toLowerCase(locale);
``` 

**Internal processing that isn’t locale-dependent:**

*Example: S3 bucket name*

Always specify the language/country-neutral locale `Locale.ROOT`. 

```
String imgTag = &#34;IMG&#34;; 
if (&#34;img&#34;.equals(imgTag.toLowerCase(Locale.ROOT)) {    
// do something 
}
```